src/Events.Functions/Services/CertificateResolverService.cs

```diff
@@ -0,0 +1,75 @@
+using Altinn.Platform.Events.Functions.Configuration;
+using Altinn.Platform.Events.Functions.Services.Interfaces;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.Platform.Events.Functions.Services
+{
+    /// <summary>
+    /// Class to resolve certificate to generate an access token
+    /// </summary>
+    /// <remarks>This class is excluded from code coverage because it has no logic to be tested.</remarks>
+    [ExcludeFromCodeCoverage]
+    public class CertificateResolverService : ICertificateResolverService
+    {
+        private readonly ILogger<ICertificateResolverService> _logger;
+        private readonly CertificateResolverSettings _certificateResolverSettings;
+        private readonly IKeyVaultService _keyVaultService;
+        private readonly KeyVaultSettings _keyVaultSettings;
+        private DateTime _expiryTime;
+        private X509Certificate2 _cachedX509Certificate = null;
+        private readonly object _lockObject = new object();
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        /// <param name="logger">The logger</param>
+        /// <param name="certificateResolverSettings">Settings for certificate resolver</param>
+        /// <param name="keyVaultService">Key vault service</param>
+        /// <param name="keyVaultSettings">Key vault settings</param>
+        public CertificateResolverService(
+            ILogger<ICertificateResolverService> logger,
+            IOptions<CertificateResolverSettings> certificateResolverSettings,
+            IKeyVaultService keyVaultService,
+            IOptions<KeyVaultSettings> keyVaultSettings)
+        {
+            _logger = logger;
+            _certificateResolverSettings = certificateResolverSettings.Value;
+            _keyVaultService = keyVaultService;
+            _keyVaultSettings = keyVaultSettings.Value;
+            _expiryTime = DateTime.MinValue;
+        }
+
+        /// <summary>
+        /// Find the configured 
+        /// </summary>
+        /// <returns></returns>
+        public async Task<X509Certificate2> GetCertificateAsync()
+        {
+            if (_expiryTime > DateTime.UtcNow && _cachedX509Certificate != null)
+            {
+                return _cachedX509Certificate;
+            }
+
+            string certBase64 = await _keyVaultService.GetCertificateAsync(
+                _keyVaultSettings.KeyVaultURI,
+                _keyVaultSettings.PlatformCertSecretId);
+
+            lock (_lockObject)
+            {
+                _cachedX509Certificate = new X509Certificate2(
+                    Convert.FromBase64String(certBase64),
+                    (string)null,
+                    X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.Exportable);
+                _expiryTime = DateTime.UtcNow.AddSeconds(_certificateResolverSettings.CacheCertLifetimeInSeconds - 5); // Set the expiry time to one hour from now
```

Comment:
This `- 5` is unnecessary. The certificate doesn't actually expire. The thing that expire is how long we hold on to the instance before reloading the same certificate. Maybe we should call the time something else? Something "reload"?